src/main/java/teammates/storage/api/ProfilesDb.java

```diff
@@ -219,50 +222,60 @@ private StudentProfile getStudentProfileEntityForLegacyData(String googleId) {
     private StudentProfile getStudentProfileEntityFromDb(String googleId) {
-        Key childKey = KeyFactory.createKey(Account.class.getSimpleName(), googleId)
-                                 .getChild(StudentProfile.class.getSimpleName(), googleId);
-
-        try {
-            StudentProfile profile = getPm().getObjectById(StudentProfile.class, childKey);
-            if (profile == null
-                    || JDOHelper.isDeleted(profile)) {
-                return null;
-            }
+        Key<Account> parentKey = Key.create(Account.class, googleId);
+        Key<StudentProfile> childKey = Key.create(parentKey, StudentProfile.class, googleId);
+        StudentProfile profile = ofy().load().key(childKey).now();
 
-            return profile;
-        } catch (JDOObjectNotFoundException je) {
+        if (profile == null) {
             return getStudentProfileEntityForLegacyData(googleId);
         }
+
+        return profile;
     }
 
     @Override
-    protected Object getEntity(EntityAttributes attributes) {
+    protected LoadType<StudentProfile> load() {
+        return ofy().load().type(StudentProfile.class);
+    }
+
+    @Override
+    protected StudentProfile getEntity(StudentProfileAttributes attributes) {
         // this method is never used and is here only for future expansion and completeness
-        return getStudentProfileEntityFromDb(((StudentProfileAttributes) attributes).googleId);
+        return getStudentProfileEntityFromDb(attributes.googleId);
     }
 
     @Override
-    protected QueryWithParams getEntityKeyOnlyQuery(EntityAttributes attributes) {
-        Class<?> entityClass = StudentProfile.class;
-        String primaryKeyName = StudentProfile.PRIMARY_KEY_NAME;
-        StudentProfileAttributes spa = (StudentProfileAttributes) attributes;
-        String id = spa.googleId;
+    protected QueryKeys<StudentProfile> getEntityQueryKeys(StudentProfileAttributes attributes) {
+        Key<Account> parentKey = Key.create(Account.class, attributes.googleId);
+        Key<StudentProfile> childKey = Key.create(parentKey, StudentProfile.class, attributes.googleId);
+        return load().filterKey(childKey).keys();
+    }
 
-        Query q = getPm().newQuery(entityClass);
-        q.declareParameters("String idParam");
-        q.setFilter(primaryKeyName + " == idParam");
+    private QueryKeys<StudentProfile> getEntityQueryKeysForLegacyData(StudentProfileAttributes attributes) {
+        Key<StudentProfile> legacyKey = Key.create(StudentProfile.class, attributes.googleId);
+        return load().filterKey(legacyKey).keys();
+    }
 
-        return new QueryWithParams(q, new Object[] {id}, primaryKeyName);
+    @Override
+    public boolean hasEntity(StudentProfileAttributes attributes) {
+        if (getEntityQueryKeys(attributes).first().now() != null) {
+            return true;
+        }
+        return getEntityQueryKeysForLegacyData(attributes).first().now() != null;
```

Comment:
Flip in what way? We should check for legacy data only if the main check fails, as legacy data is far less common.